### PR TITLE
Port TensorFlow T5 ranker to PyTorch

### PR DESCRIPTION
**Pull request by daemon**
- Implement T5 checkpoint loader
- Implement greedy decoder
- Update sample environment to use checkpoint URL

### CODE REVIEW

File: api/requirements.txt [hunk 1]
```diff
@@ -7,4 +7,4 @@ spacy==2.2.4
 tensorflow-gpu==2.1.0
 tensorflow-text==2.1.1
```
Comment (rodrigonogueira4) [inline api/requirements.txt:8]: We can remove these two now

File: api/app/modeling/serialize.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+from tensorflow.python.lib.io import file_io
+from transformers import T5Config, T5ForConditionalGeneration
+
+
+__all__ = ['CachedT5ModelLoader']
+TRANSFO_PREFIX = 'model.ckpt'
+
+
+@dataclass
+class CachedT5ModelLoader:
```
Comment (rodrigonogueira4) [inline api/app/modeling/serialize.py:17]: Can we just use whatever huggingface implements as checkpoint caching?

File: api/requirements.txt (hunk 1 above)
Comment (daemon) [reply api/requirements.txt:8]: We still need TF for `from tensorflow.python.lib.io import file_io` -- maybe we can switch to the CPU version? I also briefly tried the Google Cloud storage Python library before, but it required credentials while the TF implementation allows anonymous accounts.

Comment (rodrigonogueira4) [reply api/requirements.txt:8]: I see.. but do we still need tensorflow-text? It was used by T5's Sentencepiece tokenizer, but I believe huggingface uses its own implementation.

**Comment by edwinzhng** [comment]
Is the inference speed about the same as the TF model?

File: api/requirements.txt (hunk 1 above)
Comment (daemon) [reply api/requirements.txt:8]: Nah, we don't need that.

File: api/app/modeling/serialize.py (hunk 2 above)
Comment (daemon) [reply api/app/modeling/serialize.py:17]: Unfortunately, HF supports neither GCP downloading nor the creation of custom checkpoints. Also, I plan to refactor this module in the future once we have other caching needs.

Comment (daemon) [reply api/app/modeling/serialize.py:17]: I tried to keep the implementation as close in spirit as possible to HF's:
https://github.com/huggingface/transformers/blob/master/src/transformers/file_utils.py#L70

Comment (rodrigonogueira4) [reply api/app/modeling/serialize.py:17]: Cool, thanks for the clarification!

**Comment by daemon** [comment]
I can't measure the GPU performance on my dev machine, since the TF implementation throws OOM (I think it always pads to a sequence length of 256 and batch size of 96?). I have plenty of main RAM, so I ran both implementations on the CPU:

batch size 96, sequence length 256 (5 iterations):
PyTorch: 171.86 seconds, TensorFlow: 179 secs

Obviously, it's much faster with a smaller batch size (does that happen often in practice?):
batch size 1, sequence length 256 (5 iterations):
PyTorch: 4 seconds, TensorFlow: 179 secs

**Comment by daemon** [comment]
Oh, it doesn't seem like I have write access to merge the branch.

**Comment by rodrigonogueira4** [comment]
> Oh, it doesn't seem like I have write access to merge the branch.

Could you please try again?

**Pull request merged**